### PR TITLE
[Flaky Tests] Make PersistentTopicE2ETest#testBrokerConnectionStats tests stable.

### LINKED ISSUE #10150: Flaky-test: PersistentTopicE2ETest.testBrokerConnectionStats / PersistentTopicStreamingDispatcherE2ETest.testBrokerConnectionStats
Flaky-test: org.apache.pulsar.broker.service.PersistentTopicE2ETest.testBrokerConnectionStats
Number of failures: 12

org.apache.pulsar.broker.service.PersistentTopicE2ETest is flaky. The testBrokerConnectionStats test method fails sporadically.

```
java.lang.AssertionError: expected [0] but found [1]
        at org.testng.Assert.fail(Assert.java:99)
        at org.testng.Assert.failNotEquals(Assert.java:1037)
        at org.testng.Assert.assertEqualsImpl(Assert.java:140)
        at org.testng.Assert.assertEquals(Assert.java:122)
        at org.testng.Assert.assertEquals(Assert.java:797)
        at org.testng.Assert.assertEquals(Assert.java:807)
        at org.apache.pulsar.broker.service.PersistentTopicE2ETest.testBrokerConnectionStats(PersistentTopicE2ETest.java:1506)
```

Usage tip: To enable automatic navigation to failure message, open the following links with CTRL/CMD-click.
[example failure 2021-04-03T11:15:23.6234257Z](https://github.com/apache/pulsar/runs/2259311127?check_suite_focus=true#step:8:624)
[example failure 2021-04-02T16:29:14.4813072Z](https://github.com/apache/pulsar/runs/2254561747?check_suite_focus=true#step:8:1387)
[example failure 2021-04-02T15:36:51.8947029Z](https://github.com/apache/pulsar/runs/2254571277?check_suite_focus=true#step:8:600)
[example failure 2021-04-02T10:01:02.9570249Z](https://github.com/apache/pulsar/runs/2252764521?check_suite_focus=true#step:8:580)
[example failure 2021-04-02T09:57:27.1343835Z](https://github.com/apache/pulsar/runs/2252755426?check_suite_focus=true#step:8:600)
[example failure 2021-04-02T04:21:09.6381132Z](https://github.com/apache/pulsar/runs/2251296363?check_suite_focus=true#step:8:580)
[example failure 2021-04-02T04:11:23.4677940Z](https://github.com/apache/pulsar/runs/2251244220?check_suite_focus=true#step:8:560)
[example failure 2021-04-02T01:11:54.9818904Z](https://github.com/apache/pulsar/runs/2250236865?check_suite_focus=true#step:8:1406)
[example failure 2021-04-01T15:35:36.9708663Z](https://github.com/apache/pulsar/runs/2245689926?check_suite_focus=true#step:8:579)
[example failure 2021-03-31T14:12:28.6718791Z](https://github.com/apache/pulsar/runs/2235472463?check_suite_focus=true#step:8:446)
[example failure 2021-03-30T14:58:20.6333280Z](https://github.com/apache/pulsar/runs/2227255661?check_suite_focus=true#step:8:466)
[example failure 2021-03-30T06:00:38.9159705Z](https://github.com/apache/pulsar/runs/2224149852?check_suite_focus=true#step:8:446)

<details>
<summary>Full exception stacktrace</summary>
<code><pre>
java.lang.AssertionError: expected [0] but found [1]
        at org.testng.Assert.fail(Assert.java:99)
        at org.testng.Assert.failNotEquals(Assert.java:1037)
        at org.testng.Assert.assertEqualsImpl(Assert.java:140)
        at org.testng.Assert.assertEquals(Assert.java:122)
        at org.testng.Assert.assertEquals(Assert.java:797)
        at org.testng.Assert.assertEquals(Assert.java:807)
        at org.apache.pulsar.broker.service.PersistentTopicE2ETest.testBrokerConnectionStats(PersistentTopicE2ETest.java:1506)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:132)
        at org.testng.internal.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:45)
        at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:73)
        at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:11)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)

</pre></code>
</details>

---

Flaky-test: org.apache.pulsar.broker.service.persistent.PersistentTopicStreamingDispatcherE2ETest.testBrokerConnectionStats
Number of failures: 20

org.apache.pulsar.broker.service.persistent.PersistentTopicStreamingDispatcherE2ETest is flaky. The testBrokerConnectionStats test method fails sporadically.

```
java.lang.AssertionError: expected [0] but found [1]
        at org.testng.Assert.fail(Assert.java:99)
        at org.testng.Assert.failNotEquals(Assert.java:1037)
        at org.testng.Assert.assertEqualsImpl(Assert.java:140)
        at org.testng.Assert.assertEquals(Assert.java:122)
        at org.testng.Assert.assertEquals(Assert.java:797)
        at org.testng.Assert.assertEquals(Assert.java:807)
        at org.apache.pulsar.broker.service.PersistentTopicE2ETest.testBrokerConnectionStats(PersistentTopicE2ETest.java:1506)
```

Usage tip: To enable automatic navigation to failure message, open the following links with CTRL/CMD-click.
[example failure 2021-04-02T21:38:57.9191834Z](https://github.com/apache/pulsar/runs/2256368447?check_suite_focus=true#step:8:975)
[example failure 2021-04-02T17:58:56.5823267Z](https://github.com/apache/pulsar/runs/2255317086?check_suite_focus=true#step:8:614)
[example failure 2021-04-02T15:39:57.3252792Z](https://github.com/apache/pulsar/runs/2254561747?check_suite_focus=true#step:8:574)
[example failure 2021-04-02T12:50:47.4128805Z](https://github.com/apache/pulsar/runs/2253611608?check_suite_focus=true#step:8:614)
[example failure 2021-04-02T10:52:12.3570456Z](https://github.com/apache/pulsar/runs/2252764753?check_suite_focus=true#step:8:1353)
[example failure 2021-04-02T10:04:27.9558184Z](https://github.com/apache/pulsar/runs/2252764521?check_suite_focus=true#step:8:616)
[example failure 2021-04-02T08:23:36.8542461Z](https://github.com/apache/pulsar/runs/2252064758?check_suite_focus=true#step:8:1397)
[example failure 2021-04-02T07:36:20.9691253Z](https://github.com/apache/pulsar/runs/2252036527?check_suite_focus=true#step:8:574)
[example failure 2021-04-02T05:04:50.9540054Z](https://github.com/apache/pulsar/runs/2251244220?check_suite_focus=true#step:8:1355)
[example failure 2021-04-02T00:13:04.4592528Z](https://github.com/apache/pulsar/runs/2249631010?check_suite_focus=true#step:8:1350)
[example failure 2021-04-01T20:07:41.5538938Z](https://github.com/apache/pulsar/runs/2247059533?check_suite_focus=true#step:8:1398)
[example failure 2021-04-01T17:51:26.2247828Z](https://github.com/apache/pulsar/runs/2246206809?check_suite_focus=true#step:8:1387)
[example failure 2021-04-01T11:53:52.1246226Z](https://github.com/apache/pulsar/runs/2244635194?check_suite_focus=true#step:8:590)
[example failure 2021-04-01T11:03:04.0119695Z](https://github.com/apache/pulsar/runs/2244414606?check_suite_focus=true#step:8:1397)
[example failure 2021-04-01T05:17:24.3980445Z](https://github.com/apache/pulsar/runs/2242797184?check_suite_focus=true#step:8:1314)
[example failure 2021-03-31T21:03:46.2161428Z](https://github.com/apache/pulsar/runs/2238496507?check_suite_focus=true#step:8:1325)
[example failure 2021-03-31T20:46:37.4574830Z](https://github.com/apache/pulsar/runs/2240185745?check_suite_focus=true#step:8:568)
[example failure 2021-03-31T15:18:26.3112968Z](https://github.com/apache/pulsar/runs/2235472463?check_suite_focus=true#step:8:1349)
[example failure 2021-03-31T09:28:48.1295553Z](https://github.com/apache/pulsar/runs/2234597515?check_suite_focus=true#step:8:588)
[example failure 2021-03-29T17:25:55.2115982Z](https://github.com/apache/pulsar/runs/2219302689?check_suite_focus=true#step:8:590)

<details>
<summary>Full exception stacktrace</summary>
<code><pre>
java.lang.AssertionError: expected [0] but found [1]
        at org.testng.Assert.fail(Assert.java:99)
        at org.testng.Assert.failNotEquals(Assert.java:1037)
        at org.testng.Assert.assertEqualsImpl(Assert.java:140)
        at org.testng.Assert.assertEquals(Assert.java:122)
        at org.testng.Assert.assertEquals(Assert.java:797)
        at org.testng.Assert.assertEquals(Assert.java:807)
        at org.apache.pulsar.broker.service.PersistentTopicE2ETest.testBrokerConnectionStats(PersistentTopicE2ETest.java:1506)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:132)
        at org.testng.internal.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:45)
        at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:73)
        at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:11)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)

</pre></code>
</details>

### PR DESCRIPTION
**Pull request by horizonzy**
Fixes #10150 

### Motivation
Make PersistentTopicE2ETest#testBrokerConnectionStats tests stable.

Need to update docs? r)
  
- [x] `no-need-doc` 
(Please explain why)

### CODE REVIEW

**Review by hangc0276 — APPROVED** [review]
Good job!

**Review by Technoboy- — APPROVED** [review]

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/PulsarChannelInitializer.java [hunk 1]
```diff
@@ -59,7 +59,8 @@ public class PulsarChannelInitializer extends ChannelInitializer<SocketChannel>
     // This cache is used to maintain a list of active connections to iterate over them
     // We keep weak references to have the cache to be auto cleaned up when the connections
     // objects are GCed.
-    private final Cache<SocketAddress, ServerCnx> connections = Caffeine.newBuilder()
+    // protected for test.
```
Comment (nicoloboschi) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/PulsarChannelInitializer.java:62]: you can add @VisibleForTesting annotation

**Review by nicoloboschi — APPROVED** [review]

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/PulsarChannelInitializer.java (hunk 1 above)
Comment (horizonzy) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/PulsarChannelInitializer.java:62]: fine

**Pull request merged**